### PR TITLE
Feature/project models

### PR DESCRIPTION
**Pull request by ixboy**
Hello, This is the Private Events project. The idea is to create an app where the members of the website can publish events, attend events and invite other users, I have added all the features as required. Please review the project and inform me if there are any changes needed to be made.

Thank You.

### CODE REVIEW

**Review by cyonii — COMMENTED** [review]
.

File: db/migrate/20210403001933_create_events.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class CreateEvents < ActiveRecord::Migration[6.1]
+  def change
+    create_table :events do |t|
+      t.string :event_name
+      t.date :event_date
+      t.text :description
+      t.references :user, foreign_key: 'creator_id'
+
+      t.timestamps
```
Comment (cyonii) [inline db/migrate/20210403001933_create_events.rb:9]: Add index for `creator_id` column as in below. Remember that you do not edit your migration files by hand. You should add the index by creating a new migration.

```ruby
# Your database schema should like 

def change
  create_table :events do |t|
    t.string :event_name
    # ...rest
  end

  add_index :events, :creator_id
end
```

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class UsersController < ApplicationController
+  def index; end
+
+  def show
+    @user = User.find(params[:id])
+
+    my_events = Event.where(creator: Current.user.id)
```
Comment (cyonii) [inline app/controllers/users_controller.rb]:
```ruby

my_events = Current.user.events
```

**Review by cyonii — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️✍️

Hello @ixboy, you have done a great job in this project. I applaud you so greatly for your effort. You manage to cover most the mandatory requirements in your first CRR. That's a lot 👏.

## Highlights
- Your commit messages are great
- Nice PR description
-  Great job with your authentication. Kudos for going the extra mile of adding proper authentication 👏
- Good use of git-flow
- Clean code base
- Not a single logic in your view 👏

## To Improve
- [x] While following the step-by-step instruction on your README I was hit with some `missing gems` error, so I had to run `bundle` first. Your README is missing the instruction to install your app's dependencies. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/descriptive_readme.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
- [x] In the User show view, I should be able to see upcoming events previously attended by user using event model scope, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [ ] In your `Invitation` table, the `event_id` and `user_id` columns should be indexed
> Why indices for foreign keys are necessary? In short:
Since the foreign key columns will be used often to retrieve related information, it's important to index them (queries on those columns will be faster with indexes).
More details:
> https://semaphoreci.com/blog/2017/05/09/faster-rails-is-your-database-properly-indexed.html
> https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys

#### Tests

- [ ] Models should have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- [ ] Authentication should have have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- [ ] Event managment features should have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

## Nice to have (Optional)
- [ ] In the event show page, it would be nice if the displayed creator username is a permalink to their. show page.

Have fun coding

> #### <br>You can ping me on Slack @[CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) to draw my attention in case I do not respond quickly on here. I'm always available and will be glad to help.<br><br>

File: db/migrate/20210403001933_create_events.rb (hunk 1 above)
Comment (ixboy) [reply db/migrate/20210403001933_create_events.rb:9]: Hello, Thank you a lot for your review, but I do not have a **creator_id** column  in my Event database, I do have a "user_id" it does the work of a creator_id, and it already has an index, I don't see why should I add more unnecessary column that I will not need to use it.

Please explain to me more if I'm wrong.

Comment (cyonii) [reply db/migrate/20210403001933_create_events.rb:9]: I guess I was being specific on the project's specifications which say to add a `creator_id` in your `events` table.

In the code snippet below, I would like to understand what you intend to achieve with this part `foreign_key: 'creator_id'`, it seems strange, I could learn.
```ruby
t.references :user, foreign_key: 'creator_id'
```

Comment (ixboy) [reply db/migrate/20210403001933_create_events.rb:9]: Hello Cyonii, good observation thanks for taking your time to reply to me :), I believe that was just me playing around and learning how to implement migrations but I later changed that code and did not use it anymore, if you look in my schema.rb the events table does not have that.

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :red_circle: :stop_sign: 
-------------
Hello @ixboy You did an excellent job working on the required changes. :clap: 
However, there are still some changes to make :hammer_and_wrench: see them below :point_down: :eyes: 
---------
- [ ] You still missing the test asked for the previous TSE. I can see that you write the tests on the folder test. Those are not the tests that you must work on. You have to work with the RSpec and Capybara to write the correct tests.  Please work on that. 
- [ ]  Models should have unit tests for validations and associations (Rspec) 
- [ ] Authentication should have have integration tests (Rspec + Capybara) 
- [ ] Event managment features should have integration tests (Rspec + Capybara)
------------
Make the required changes and send it again for another review. :1st_place_medal: :muscle: 
Reach me on SLACK if you have any questions.
@ Brian Sammit Cruz Rodriguez
HAPPY CODING!!! :smile:

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Ismael Ngalula Antonio** 
Good job with the improvements for this project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
Just one last detail to improve before you can go ahead:

- [ ] In your tests, use only one "expect" for each "it"
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by voscarmv — APPROVED** [review]
# Project status: Approved 💯 ✨ 😎 👍 
### You may merge this branch and move forward ⛙🛣️ 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻
# Happy Coding! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**